# QRT ENS Data Challenge 2023 - Benchmark (english)

This notebook details the benchmark construction for the challenge - it may also help participants to start the competition.

## Libraries

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import spearmanr
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

## Loading data

- `X_train` and `X_test` both have $35$ columns that represent the same explanatory variables but over different time periods. 

- `X_train` and `Y_train` share the same column `ID` - each row corresponds to a unique ID associated wwith a day and a country. 

- The target of this challenge `TARGET` in `Y_train` corresponds to the price change for daily futures contracts of 24H electricity baseload. 

- **You will notice some columns have missing values**.


In [2]:
# After downloading the X_train/X_test/Y_train .csv files in your working directory:
data_path = '../data/'
X_train = pd.read_csv(data_path+'X_train.csv')
Y_train = pd.read_csv(data_path+'Y_train.csv')
X_test = pd.read_csv(data_path+'X_test.csv')

In [3]:
X_train.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1054     206      FR        0.210099       -0.427458       -0.606523   
1  2049     501      FR       -0.022399       -1.003452       -0.022063   
2  1924     687      FR        1.395035        1.978665        1.021305   
3   297     720      DE       -0.983324       -0.849198       -0.839586   
4  1101     818      FR        0.143807       -0.617038       -0.924990   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.606523            NaN       0.692860            NaN  ...   
1        0.022063      -0.573520      -1.130838       0.573520  ...   
2       -1.021305      -0.622021      -1.682587       0.622021  ...   
3        0.839586      -0.270870       0.563230       0.270870  ...   
4        0.924990            NaN       0.990324            NaN  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.444661 -0.172680 -0.556356 -0.790823 -0.283160 -1.069070   
1         -1.183194 -1.240300 -0.770457  1.522331  0.828412  0.437419   
2          1.947273 -0.480700 -0.313338  0.431134  0.487608  0.684884   
3         -0.976974 -1.114838 -0.507570 -0.499409 -0.236249  0.350938   
4         -0.526267 -0.541465 -0.424550 -1.088158 -1.011560  0.614338   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0 -0.063404  0.339041  0.124552   -0.002445  
1  1.831241 -0.659091  0.047114   -0.490365  
2  0.114836  0.535974  0.743338    0.204952  
3 -0.417514  0.911652 -0.296168    1.073948  
4  0.729495  0.245109  1.526606    2.614378  

[5 rows x 35 columns]

In [4]:
Y_train.head()

ID    TARGET
0  1054  0.028313
1  2049 -0.112516
2  1924 -0.180840
3   297 -0.260356
4  1101 -0.071733

## Model and train score

The benchark for this challenge consists in a simple linear regression, after a light cleaning of the data: The missing (NaN) values are simply filled with 0's and the `COUNTRY` column is dropped - namely we used the same model for France and Germany. 

In [23]:
X_train

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0     1054     206      FR        0.210099       -0.427458       -0.606523   
1     2049     501      FR       -0.022399       -1.003452       -0.022063   
2     1924     687      FR        1.395035        1.978665        1.021305   
3      297     720      DE       -0.983324       -0.849198       -0.839586   
4     1101     818      FR        0.143807       -0.617038       -0.924990   
...    ...     ...     ...             ...             ...             ...   
1489   459     809      DE        1.529204        1.106682       -1.855327   
1490  1674     887      FR        1.618582        1.752840        0.611392   
1491   748    1083      DE        0.856399        0.489199       -0.255778   
1492  1454    1133      FR        0.560689       -0.343777       -0.830239   
1493   359    1118      DE        0.226730       -0.711005        0.873016   

      FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0           0.606523            NaN       0.692860            NaN  ...   
1           0.022063      -0.573520      -1.130838       0.573520  ...   
2          -1.021305      -0.622021      -1.682587       0.622021  ...   
3           0.839586      -0.270870       0.563230       0.270870  ...   
4           0.924990            NaN       0.990324            NaN  ...   
...              ...            ...            ...            ...  ...   
1489        1.855327      -0.218658       1.450426       0.218658  ...   
1490       -0.611392       0.449153      -0.152146      -0.449153  ...   
1491        0.255778      -1.531544      -0.829568       1.531544  ...   
1492        0.830239      -0.304856       1.210230       0.304856  ...   
1493       -0.873016       0.296651      -1.547075      -0.296651  ...   

      FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0            -0.444661 -0.172680 -0.556356 -0.790823 -0.283160 -1.069070   
1            -1.183194 -1.240300 -0.770457  1.522331  0.828412  0.437419   
2             1.947273 -0.480700 -0.313338  0.431134  0.487608  0.684884   
3            -0.976974 -1.114838 -0.507570 -0.499409 -0.236249  0.350938   
4            -0.526267 -0.541465 -0.424550 -1.088158 -1.011560  0.614338   
...                ...       ...       ...       ...       ...       ...   
1489          0.509514       NaN       NaN       NaN       NaN       NaN   
1490          1.666252       NaN       NaN       NaN       NaN       NaN   
1491          0.358120  0.207905  0.404763 -0.594595  0.894011  0.256338   
1492         -0.184862 -0.682815 -0.390304 -0.972088 -1.501930  1.215528   
1493         -0.634501 -0.757347 -0.744355 -0.482286 -0.941070 -0.120703   

       FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0    -0.063404  0.339041  0.124552   -0.002445  
1     1.831241 -0.659091  0.047114   -0.490365  
2     0.114836  0.535974  0.743338    0.204952  
3    -0.417514  0.911652 -0.296168    1.073948  
4     0.729495  0.245109  1.526606    2.614378  
...        ...       ...       ...         ...  
1489       NaN  0.876984  0.819520    1.320373  
1490       NaN  0.932633 -0.085690    0.356356  
1491  0.402316 -1.112899 -0.237835    0.067152  
1492  1.338708  0.962812 -5.392852   -0.843812  
1493  0.391491  1.059829 -0.326640   -0.566724  

[1494 rows x 35 columns]

In [5]:
lr = LinearRegression()

X_train_clean = X_train.drop(['COUNTRY'], axis=1).fillna(0)
Y_train_clean = Y_train['TARGET']

lr.fit(X_train_clean, Y_train_clean)

output_train = lr.predict(X_train_clean)

def metric_train(output):
    return  spearmanr(output, Y_train_clean).correlation

print('Spearman correlation for the train set: {:.1f}%'.format(100 * metric_train(output_train) ))

Spearman correlation for the train set: 27.9%


The Spearman correlation obtained with our model on the train data set is about 27.9%.

NB: Electricity price variations can be quite volatile and this is why we have chosen the Spearman rank correlation as a robust metric for the challenge, instead of the more standard Pearson correlation.

ADDED MANUALY

In [21]:
import plotly.express as px

Y_train_predicted = pd.DataFrame(data=output_train, columns=['TARGET_predicted'])

In [22]:
Y_train

ID    TARGET
0     1054  0.028313
1     2049 -0.112516
2     1924 -0.180840
3      297 -0.260356
4     1101 -0.071733
...    ...       ...
1489   459 -0.172597
1490  1674 -0.063546
1491   748  0.151797
1492  1454 -0.640917
1493   359 -0.252764

[1494 rows x 2 columns]

In [ ]:
figure = px.bar()

In [6]:
figure = px.scatter(x=Y_train_clean, y=output_train)
figure.show()

In [7]:
Y_train_clean

0       0.028313
1      -0.112516
2      -0.180840
3      -0.260356
4      -0.071733
          ...   
1489   -0.172597
1490   -0.063546
1491    0.151797
1492   -0.640917
1493   -0.252764
Name: TARGET, Length: 1494, dtype: float64

## Generate the benchmark output

Next, we process the test set the same way as we did on the train set and predict using our linear model, while saving the predictions to a csv file satisfying the challenge output contraints.


In [8]:
X_test.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1115     241      FR        0.340083       -0.433604       -0.423521   
1  1202    1214      FR        0.803209        0.780411        0.601610   
2  1194    1047      FR        0.795540        0.721954        1.179158   
3  1084    1139      FR        0.172555       -0.723427       -0.044539   
4  1135     842      FR        0.949714        0.420236        0.617391   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.423521       0.165333       0.519419      -0.165333  ...   
1       -0.601610       0.342802       0.555367      -0.342802  ...   
2       -1.179158       1.620928       0.666901      -1.620928  ...   
3        0.044539            NaN      -0.205276            NaN  ...   
4       -0.617391       0.608561      -0.240856      -0.608561  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.222525 -0.513180 -0.182048 -0.982546 -0.876632  0.880491   
1          0.857739 -0.340595 -0.301094 -0.759816 -1.221443 -0.616617   
2          0.447967  0.796475 -0.367248  0.376055 -0.483363  0.865138   
3         -0.561295 -0.542606 -0.013291 -0.791119 -0.894309  0.239153   
4          0.503567 -0.230291 -0.609203 -0.744986 -1.196282  0.176557   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0  0.692242  0.569419 -0.029697   -0.929256  
1 -0.737496  0.251251  0.753646    0.664086  
2  0.120079 -1.485642 -0.326450   -0.349747  
3  0.457457 -0.746863  2.262654    0.642069  
4  0.312557 -2.219626 -0.509272   -0.488341  

[5 rows x 35 columns]

In [9]:
X_test_clean = X_test.drop(['COUNTRY'], axis=1).fillna(0)

Y_test_submission = X_test[['ID']].copy()
Y_test_submission['TARGET'] = lr.predict(X_test_clean)

def metric_train(output):
    return  spearmanr(output, Y__clean).correlation

print('Spearman correlation for the test set: {:.1f}%'.format(100 * metric_train(Y_test_submission) ))

# Y_test_submission.to_csv('benchmark_qrt.csv', index=False)


After submitting this benchmark_qrt.csv file at https://challengedata.ens.fr, we obtain a public score of around 15.9 % - this is the score to beat!


## Tips and ideas for improvements 

- Thinking about the fundamental price drivers of electricity in each country may be useful. 

- The dataset is small, so you will need to be careful not to overfit your models. It will be useful to use good crossvalidation practices. 
